Fix async decode

harmony-ek opened the pull request.

Commit:
Fix a copy-and-paste glitch

Commit:
Clarify asynchronous nature of Decode()

Commit:
Add asynchronous block notification interface

To be implemented in later commits.

Commit:
Implement common ready-block channel logic

This is to be used in Decoder implementations.

Commit:
Make SymbolSize and NumSourceBlocks independent

That is, do not get them from the wrapped encoder but calculate and
return them using common and scheme-specific OTI.

Commit:
Depend on RFC6330__v1, not on RaptorQ__v1

Commit:
Add swig.WaitForBlock(dec, sbn, err)

Make it wait for any block to be available and return its status.

FTR, potential error codes from compute() include:

* swig.Error_NONE -> source block (sbn) is ready
* swig.Error_EXITING -> underlying decoder has been deleted
* swig.Error_NEED_DATA -> source block (sbn) has reached end-of-input
  (as indicated by the caller) but still cannot be decoded.

Since we do not inject end-of-input, we will not get and do not have to
process NEED_DATA.

Commit:
Reset dec.wrapper to nil before deleting dtor

This is so that accidental references from routines that execute as a
result do not cause segmentation fault.  (We are about to add such a
code path.)

Commit:
Unroll one-shot decoder struct initialization

Because we are about to add a ready-block channel notifier as a data
member, which cannot be value-initialized in a sensible manner.

Commit:
Use ReadyBlockChannels from Decoder

chaosma commented on pkg/internal/readyblockchan/readyblockchan.go:
```diff
@@ -0,0 +1,94 @@
+type AlreadyAdded chan<- uint8
+
+func (e AlreadyAdded) Error() string {
+	return fmt.Sprintf("ready-block channel %+v already added")
```
It seems that %+v not work here.

harmony-ek replied:
Good catch, will fix.

chaosma commented on pkg/internal/readyblockchan/readyblockchan.go:
```diff
@@ -0,0 +1,94 @@
+type NotFound chan<- uint8
+
+func (e NotFound) Error() string {
+	return fmt.Sprintf("ready-block channel %+v not found")
```
%+v not work here

harmony-ek replied:
Good catch, will fix (2).

chaosma reviewed (commented):
Looks good. Some minor changes may needed

Commit:
Move internal out of pkg

This is so that outside users do not try and use ReadyBlockChan in their
code.

Commit:
Make impl internal too

They are not supposed to be directly called by outside users.

Commit:
go fmt

Commit:
goimports

Commit:
Update JetBrains IDE .gitignore

Commit:
.gitignore .idea/*.iml

Commit:
Run goimports on save from GoLand

Commit:
Add New{En,De}coder convenience functions

They use the default factory and accept the same arguments as the New
function of the factory interface does.

Commit:
Add missing self

Spotted-by: Chao Ma <chao@harmony.one>

Commit:
golint

Commit:
Move NewEncoder/NewDecoder into defaults

This is to break an import cycle.
